src/linear_algebra/general_linear_group.lean

```diff
@@ -0,0 +1,191 @@
+  ext1, dsimp at *, ext1, solve_by_elim,
+end
+
+@[ext] lemma ext (A B : GL n R) : (∀ i j, A i j = B i j) → A = B :=
+(ext_iff A B).mpr
+
+/-- Applying `matrix.nonsing_inv` to the coercion is the same as coercing the group inverse. -/
+lemma nonsing_inv_coe_eq_coe_inv (A : GL n R) : (↑A : matrix n n R)⁻¹ = ↑(A⁻¹) :=
+begin
+  letI := A.invertible,
+  exact (inv_eq_nonsing_inv_of_invertible ↑A).symm,
+end
+
+section coe_lemmas
+
+variables (A B : GL n R)
+
+@[simp] lemma mul_val : ↑(A * B) = A ⬝ B := rfl
+
+@[simp] lemma mul_apply : ⇑(A * B) = (A ⬝ B) := rfl
+
+@[simp] lemma one_val : ↑(1 : GL n R) = (1 : matrix n n R) := rfl
+
+@[simp] lemma one_apply : ⇑(1 : GL n R) = (1 : matrix n n R) := rfl
+
+@[simp] lemma inv_val : ↑(A⁻¹) = nonsing_inv A :=
+begin
+  have:=nonsing_inv_coe_eq_coe_inv A, unfold_coes at *, rw ← this, refl,
+end
+
+@[simp] lemma inv_apply : ⇑(A⁻¹) = nonsing_inv A :=
+begin
+  have:=nonsing_inv_coe_eq_coe_inv A, unfold_coes at *, rw ← this, refl,
+end
+
+@[simp] lemma to_lin'_mul : to_lin' (A * B) = (to_lin' A).comp (to_lin' B) :=
+matrix.to_lin'_mul A B
+
+@[simp] lemma to_lin'_one : to_lin' (1 : GL n R).1 = linear_map.id :=
+matrix.to_lin'_one
+
+end coe_lemmas
+
+instance SL_to_GL : has_coe (special_linear_group n R) (GL n R) :=
+⟨λ A, ⟨↑A, ↑(A⁻¹), congr_arg coe (mul_right_inv A), congr_arg coe (mul_left_inv A)⟩⟩
+
+@[simp] lemma GL_vals (A : GL n R) : ∀ i j, A i j = A.1 i j :=
+begin
+unfold_coes, intros i j,refl,
+end
+
+lemma det_not_zero [nontrivial R] (A : GL n R) : det A ≠ 0 :=
+begin
+  have := GL.is_unit_det _ _ A, simp, by_contradiction, unfold_coes at *,
+  rw h at this, simp at *, exact this,
+end
+
+end general_linear_group
+
+namespace GL_plus
+
+variables  (n : Type u) [decidable_eq n] [fintype n] (R : Type v) [linear_ordered_comm_ring R ]
+
+lemma one_in_GL_pos : 0 < det (1 : GL n R) :=
+begin
+simp only [det_one, gt_iff_lt, general_linear_group.one_apply,zero_lt_one],
+end
+
+lemma mul_in_GL_pos (A B : GL n R) (h1 : 0 < det A) (h2 : 0 < det B) : 0 < det (A*B) :=
+begin
+  simp only [gt_iff_lt, det_mul, mul_eq_mul], apply mul_pos h1 h2,
+end
+
+lemma inv_det_pos  (A : GL n R) (h : 0 < det A) :  0 < det (A⁻¹).1 :=
+begin
+  have h0 := (GL.is_unit_det _ _ A),
+  have h1 := is_unit_nonsing_inv_det A h0,
+  have h2 := nonsing_inv_det A h0,
+  have h3 : 0 < det ⇑A*  det (⇑A)⁻¹ , by {rw mul_comm, rw h2, linarith}, unfold_coes at *,
+  convert (zero_lt_mul_left h).mp h3,
+  have := general_linear_group.nonsing_inv_coe_eq_coe_inv A,  unfold_coes  at this, simp_rw this,
+end
+
+/-- This is the subgroup of `nxn` matrices with  entries over a
+linear ordered ring and positive determinant-/
+
+ def GL_pos : subgroup (GL n R) :=
+{ carrier := {M  | 0 < det M},
+  one_mem' := one_in_GL_pos _ _ ,
+  mul_mem' := λ A B h1 h2, mul_in_GL_pos _ _ A B h1 h2,
+  inv_mem' := λ A h1, by {have:= inv_det_pos _ _ A h1, simp only [set.mem_set_of_eq] at *,
+  unfold_coes at *, convert this, }
+}
```

Comment:
Style:
```suggestion
/-- This is the subgroup of `nxn` matrices with entries over a
linear ordered ring and positive determinant. -/
def GL_pos : subgroup (GL n R) :=
{ carrier := {M  | 0 < det M},
  one_mem' := one_in_GL_pos _ _ ,
  mul_mem' := λ A B h1 h2, mul_in_GL_pos _ _ A B h1 h2,
  inv_mem' := λ A h1, begin
    have:= inv_det_pos _ _ A h1,
    simp only [set.mem_set_of_eq] at *,
    unfold_coes at *,
    convert this,
  end }
```
(no space after comment, no space before def, one tactic per line, indent the `begin / end` block)

Reply:
Fixed. In general, should it be one tactic per line? For example, do I also need to fix `inv_det_pos`?

Reply:
In general, yes

Reply:
Ok I understand. I've done this now unless they are in brackets or something similar.